File: FairPlay_Streaming_Server_SDK_5.1/Development/Key_Server_Module/Swift/Sources/src/base/parse_spc_TLLVs/parseTagClientCapabilities.swift
//
// Copyright © 2023-2024 Apple Inc. All rights reserved.
//

extension Base {
  public static func parseTagClientCapabilities(
    _ tllv: FPSServerTLLV, _ spcContainer: inout FPSServerSPCContainer
  ) throws {

    // Check that size matches expected size exactly
    try requireAction(
      tllv.value.count == base_constants.FPS_CAPABILITIES_FLAGS_LENGTH, { throw FPSStatus.paramErr }
    )

    // Entire TLLV value is the client capabilities flags
    spcContainer.spcData.clientCapabilities = tllv.value

    Log.debug(
      "Client Capabilities: 0x\(spcContainer.spcData.clientCapabilities.map { String(format: "%02x", $0) }.joined())")
  }
}

File: FairPlay_Streaming_Server_SDK_5.1 2/Development/Key_Server_Module/Rust/src/base/parse_spc_TLLVs/parseTagClientCapabilities.rs
//
// Copyright © 2023-2024 Apple Inc. All rights reserved.
//

use crate::base::base_constants::FPS_CAPABILITIES_FLAGS_LENGTH;
use crate::base::structures::base_fps_structures::Base;
use crate::base::structures::base_server_structures::{FPSServerSPCContainer, FPSServerTLLV};
use crate::requireAction;
use crate::validate::{FPSStatus, Result};

impl Base {
    pub fn parseTagClientCapabilities(tllv: &FPSServerTLLV, spcContainer: &mut FPSServerSPCContainer) -> Result<()> {
        // Check that size matches expected size exactly
        requireAction!(
            tllv.value.len() == FPS_CAPABILITIES_FLAGS_LENGTH,
            return Err(FPSStatus::paramErr)
        );

        // Entire TLLV value is the client capabilities flags
        spcContainer.spcData.clientCapabilities = tllv.value.clone();

        log::debug!(
            "Client Capabilities: 0x{}",
            hex::encode(spcContainer.spcData.clientCapabilities.as_slice())
        );

        Ok(())
    }
}
